feat(cohort): implement new cohort filters

alexkim205 opened the pull request:
## Changes

Implements new cohort filters on new cohort detail page.

Changes:

- [ ] Integrates cohort row builder with shape of data
- [ ] Populating defaults for each behavioral case
- [ ] Add FF
- [ ] Conditionally render old and new cohort filters with FF
- [ ] Error states
- [ ] And more!

https://user-images.githubusercontent.com/13460330/166833095-ef7736aa-42b1-4746-aa78-1e58ecf608b3.mov


👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

<!-- Briefly describe the steps you took. -->
<!-- Include automated tests if possible, otherwise describe the manual testing routine. -->

alexkim205 committed:
refactor cohortLogic

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-cohort-pages

alexkim205 committed:
new dragger

alexkim205 committed:
static and dynamic working

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-cohort-pages

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-cohort-pages

alexkim205 committed:
rename cohorttypetype

alexkim205 committed:
address feedback

alexkim205 committed:
console

alexkim205 committed:
consoles and fix test

alexkim205 committed:
remove unused components

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-cohort-pages

alexkim205 committed:
correct data attr

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-cohort-pages

alexkim205 committed:
address feedback

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-cohort-pages

alexkim205 committed:
fix test

alexkim205 committed:
feat: new cohort selectors

alexkim205 committed:
address feedback

alexkim205 committed:
fix test

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-cohort-menus

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-cohort-menus

alexkim205 committed:
feat: add groups to actors dropdown and logic test

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-cohort-menus

alexkim205 committed:
chore: remove in the last and initial

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/new-cohort-menus

alexkim205 committed:
rename

alexkim205 committed:
initial grammar constants

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-grammar

alexkim205 committed:
added all variations of filters

alexkim205 committed:
add all fields and stories

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-grammar

alexkim205 committed:
defaults

alexkim205 committed:
fix tests

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-grammar

alexkim205 committed:
styling

alexkim205 committed:
cohort filter implement

alexkim205 committed:
added corresponding keys to all rows

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-grammar

alexkim205 committed:
feat: add backwards and forwards conversion utils for cohort groups

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/and-or-refactor

alexkim205 committed:
keep in vcs

alexkim205 committed:
typing

alexkim205 committed:
simplify return

alexkim205 committed:
add filters to cohort page

alexkim205 committed:
remove unused util

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/and-or-refactor

alexkim205 committed:
populate with default values

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/and-or-refactor

alexkim205 committed:
remove irrelevant comment

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-filter-logic

alexkim205 committed:
validation checks

alexkim205 committed:
hide delete icons

alexkim205 committed:
add form tests

alexkim205 committed:
bad merge

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-filter-logic

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-filter-logic

alexkim205 committed:
nullify other props

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-filter-logic

alexkim205 committed:
json load issue

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-filter-logic

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-filter-logic

alexkim205 committed:
fix empty state tests

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-filter-logic

alexkim205 committed:
cohort py revert

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-filter-logic

alexkim205 committed:
errors

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-filter-logic

alexkim205 committed:
refactor error logic

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-filter-logic

alexkim205 committed:
refactor utils

alexkim205 committed:
cohort selector values show

alexkim205 committed:
edge cases

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-filter-logic

alexkim205 committed:
remove consoles

alexkim205 committed:
prettier

alexkim205 committed:
remove hardcode

alexkim205 committed:
bug fix

alexkim205 committed:
ts

alexkim205 committed:
old groups broken bug fix

alexkim205 committed:
ts

alexkim205 committed:
lint

EDsCODE reviewed (commented):
Quick UI review driveby. Will give full code review tomorrow 

Needs more space at the bottom
<img width="398" alt="Screen Shot 2022-05-05 at 9 07 32 PM" src="https://user-images.githubusercontent.com/13127476/167054297-1dc07132-3da2-4983-bcb7-50f62028db7a.png">

TODO to handle 🙃
<img width="483" alt="Screen Shot 2022-05-05 at 9 10 02 PM" src="https://user-images.githubusercontent.com/13127476/167054314-bfa4a60e-21d8-41d5-8d5c-40543b8eea2f.png">

Can't type new number in first field on "complete event multiple times"
<img width="538" alt="Screen Shot 2022-05-05 at 9 59 27 PM" src="https://user-images.githubusercontent.com/13127476/167054475-4a77dc9f-2d2c-41b9-9af6-29aeb3f845ad.png">

alexkim205 committed:
bugs and add property value selector

alexkim205 commented:
> Also quick flyby around copy: cohorts never deal with event properties, only person properties!
> 
> In the same vein, not sure if `[do not] have the property` belongs in the `behavioral` section

Yeah, I think that's the assumption I ran with since properties are grouped together with event behaviors. @clarkus would another section called "Person properties" in the dropdown be clearer here?

EDsCODE commented on frontend/src/scenes/cohorts/CohortFilters/CohortField.tsx:
```diff
@@ -15,40 +16,47 @@ import {
+        if (props.fieldKey) {
+            onChange(props.criteria)
+        }
+    }, [resolveCohortFieldValue(props.criteria, props.fieldKey)])
```
What's the reasoning for having this logic hook? Also, it seems like onChange does the exact checks that would happen here. `resolveCohortFieldValue`

alexkim205 replied:
This keeps the logic in sync with the component props, since as of now it's impossible to do that within kea logic itself. It's hacky but we do this in `ActionFilter` too - https://github.com/PostHog/posthog/blob/812efd0b1587602de3c78b3c35f9f70e39a7897d/frontend/src/scenes/insights/ActionFilter/ActionFilter.tsx#L142-L146

EDsCODE commented on frontend/src/scenes/cohorts/cohortUtils.ts:
```diff
@@ -0,0 +1,424 @@
+    }
+}
+
+export function processCohortOnSet(cohort: CohortType, isNewCohortFilterEnabled: boolean = false): CohortType {
```
I'm guessing this is tested implicitly be the cohortlogic tests. What's the general pattern/decision behind this? It seems like a lot of utils in this file could be tested independently of the cohort logic

alexkim205 replied:
There is a lot of repeating functionality used in various actions in the logic, so I refactored a bunch out of the logic and into utils. It'd be a nice to have but I don't think its too necessary to individually test these again since they're all being tested a number of times in `cohortLogic.test.ts`

EDsCODE reviewed (commented):
Pass 1

alexkim205 commented:
All addressed!

> Can't type new number in first field on "complete event multiple times" <img alt="Screen Shot 2022-05-05 at 9 59 27 PM" width="538" src="https://user-images.githubusercontent.com/13127476/167054475-4a77dc9f-2d2c-41b9-9af6-29aeb3f845ad.png">

This bug was a really sneaky one. React was reusing the same component instance in the case where two of the same field types (number fields) existed in the same criteria row since all fields were keyed by index, not by property key. Replacing `key={i} with key={fieldKey}` triggers a component rerender which also sure the correct keyed logic is being used.

alexkim205 committed:
fix rererender fields bug

alexkim205 commented on frontend/src/scenes/cohorts/Cohort.tsx:
```diff
@@ -210,86 +202,261 @@ export function Cohort({ id }: { id?: CohortType['id'] } = {}): JSX.Element {
+                                                </Group>
+                                            ))}
+                                            <span id="add" />
+                                            <div style={{ marginTop: 8, marginBottom: 8 }}>
```
There's no real standard since use both liberally everywhere. Here I just opted for inline since it wasn't worth sticking another class into the scss for just some margins.

alexkim205 commented on frontend/src/scenes/cohorts/CohortFilters/CohortCriteriaRowBuilder.tsx:
```diff
@@ -1,45 +1,172 @@
+                                            <KeaField
+                                                key={i}
+                                                name={field.fieldKey}
+                                                template={({ error, kids }) => {
```
👨‍👧‍👦

alexkim205 committed:
bottom start

alexkim205 commented on frontend/src/scenes/cohorts/Cohort.tsx:
```diff
@@ -210,86 +202,261 @@ export function Cohort({ id }: { id?: CohortType['id'] } = {}): JSX.Element {
-                                                    alignItems: 'center',
-                                                    width: '100%',
-                                                }}
+                                    <Row align="middle" justify="space-between" wrap={false}>
```
I'm not sure if splitting these down into subrender methods will be more readable since they're only being used once in this context and persisting these as variables before the return adds a lot of code in the main method that otherwise won't be rendered in certain contexts. But I see what you mean ternary operators and nested components are hard to read. 

I'll refactor this out into a separate CohortCriteriaGroup component that binds to the same cohortlogic. Hopefully this helps w readability

alexkim205 committed:
refactor group out

alexkim205 committed:
pluralize group

neilkakkar reviewed (commented):
Wow, nice stuff!

I haven't yet jumped into the code, but here's few things to take care of: (okay to do in follow ups, since behind FF, your call)

1. When going to edit view of both old and new cohorts, the values aren't filled where they should be. On save, they become empty too.
2. A suspicious `id` is being sent with parameters on save, where property filters should only have `type` and `values` keys: 
<img width="359" alt="image" src="https://user-images.githubusercontent.com/7115141/167392782-32d60641-b190-4df3-bd5a-badd75727500.png">

3. The operators for events should be limited to 3: gte, lte, or eq. The `eq` on the frontend shows up as `exact`, which cohort backend doesn't yet support. I'll fix this to include `exact` as well, since everywhere else supports that. Makes things easier for you I think?
<img width="865" alt="image" src="https://user-images.githubusercontent.com/7115141/167393509-e05e5554-fb28-483f-89ca-51dae9ca89bb.png">

4. When opening an old cohort, the first match group defaults to `AND` when sending the request, but shows up as `OR` in the UI.

5. When there are errors calculating, there's no indication on the UI? So, basically, if I am with an invalid cohort, there's no way to know, and my persons results are stale :(

6. Validation for sequence, stopped performing event, restarted performing event is missing, which silently errors out (point 5): In each case, seq_value must be more recent than date_value, or in other words, lesser:
<img width="896" alt="image" src="https://user-images.githubusercontent.com/7115141/167398153-557fc899-5964-4cc0-a955-b1d836b30cab.png">
These are valid, but if you reverse the numbers in either case, they should be invalid^^

7. ~~Negation validation seems to error out as well based on ordering~~ backend issue, will fix!

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-filter-logic

alexkim205 committed:
address feedback

alexkim205 committed:
fix logical type selector

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-filter-logic

alexkim205 commented:
Thanks for the review!

1. I'm not sure how I can reproduce this error. It seems like all the fields are being populated correctly on editing both old and new cohorts.
2. Yes good catch! I'll nix these before sending to the server
3. Thanks, I think handling `exact` on the backend makes sense to make it consistent w/ how we process property filters already.
4. fixed
5. In the situation where there are errors calculating, it would be super helpful to receive the specific error message as well so that I can surface a nongeneric error toast on the frontend. If this isn't possible, I can also just surface a generic error during calculating error. For now I've added an error toast like below:

<img width="438" alt="Screen Shot 2022-05-09 at 1 22 52 PM" src="https://user-images.githubusercontent.com/13460330/167463759-967f6d01-72ef-4a9c-8e33-61cae919ad53.png">

6. Nice, implemented

<img width="631" alt="Screen Shot 2022-05-09 at 1 40 57 PM" src="https://user-images.githubusercontent.com/13460330/167466563-dc33b513-2f65-42f7-bc0c-ee21b3e12ace.png">

alexkim205 committed:
add sequential error handling

alexkim205 committed:
primary-alt to alt

alexkim205 committed:
fix logic tests

rcmarron commented on frontend/src/scenes/cohorts/CohortFilters/constants.tsx:
```diff
@@ -497,223 +539,229 @@ export const ROWS: Record<BehavioralFilterType, Row> = {
                 type: FilterType.Number,
-                value: 5,
+                defaultValue: 5,
             },
```
I think we want 1 Text element on the end here that says `periods`

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-filter-logic

rcmarron commented on frontend/src/scenes/cohorts/CohortFilters/constants.tsx:
```diff
@@ -497,223 +539,229 @@ export const ROWS: Record<BehavioralFilterType, Row> = {
+                fieldKey: 'seq_time_value',
                 type: FilterType.Number,
-                value: 30,
+                defaultValue: '30',
```
Let's make this be 15, so it's smaller than the time value by default

rcmarron replied:
We should probably do this everywhere we set `seq_time_value`

rcmarron reviewed (commented):
Really really great work here!

A couple nits in the comments, and I noticed a few bugs:

1. Selected actions appear as their action id once they're selected instead of the action name
![image](https://user-images.githubusercontent.com/4813045/167497867-d7367c6b-8926-4922-935a-97f56fa257b7.png)
2. For criteria that use `seq_event` I don't think we're setting `seq_event_type` (this is my bad, I never covered it in the grammar doc)
3. Our negation error messaging is a bit odd. It’s not obvious what’s ‘negation criteria’ and what isn’t. I think we should say something like: “‘Did not complete a sequence of events’ is a negative cohort criteria. Negative criteria can only be used when matching all criteria (AND), and must be accompanied by at least one positive matching criteria.”
4. There's odd behavior when you change criteria type. For example, changing a criteria type from “Have property” to “In cohort” can persist the property value as the cohort name.
![image](https://user-images.githubusercontent.com/4813045/167498202-27166545-f28a-420f-b123-5dee26442d86.png)
5. There's something odd with errors in cohort calculation. I'm getting the error toast on every save. I don't know if this is a backend bug or not. Similarly, it doesn't seem like we're clearing + updating the persons in the cohort list on each save.

rcmarron commented on frontend/src/scenes/cohorts/CohortFilters/types.ts:
```diff
@@ -79,16 +90,47 @@ export interface CohortTaxonomicFieldProps extends Omit<CohortFieldBaseProps, 'f
+export enum CohortClientErrors {
+    NegationCriteriaMissingOther = 'Negation criteria are only supported after you have specified at least one positive matching criteria. Negation criteria can only be used when matching all criteria (AND).',
+    NegationCriteriaCancel = 'These criteria cancel each other out, and would result in no matching persons.',
+    RegularEventMismatch = 'The lowerbound period value must not be greater than the upperbound value.',
```
Also, `RegularEventMismatch` seems like an odd name here.

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-filter-logic

alexkim205 commented:
1. Good catch! fixed
2. No worries, fixing bug 1 also fixes this one.
3. Done
 
<img width="779" alt="Screen Shot 2022-05-10 at 12 55 28 PM" src="https://user-images.githubusercontent.com/13460330/167682221-d8b11611-e6ec-49a3-9683-fa2ec2bb335e.png">
4. Fixed
5. I can't seem to reproduce this bug. I'm trying both filter configs where there would and wouldn't be errors during calculations.

alexkim205 committed:
address feedback

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-filter-logic

alexkim205 committed:
fix tests

clarkus commented:
> Yeah, I think that's the assumption I ran with since properties are grouped together with event behaviors in the mockups. @clarkus would another section called "Person properties" in the dropdown be clearer here?

Sure I think that's clarifying and doesn't do any harm so why not. 

<img width="293" alt="Screen Shot 2022-05-10 at 2 57 52 PM" src="https://user-images.githubusercontent.com/254612/167728615-5abfdb5a-ee4e-4139-b1ee-584bea137bc8.png">

rcmarron reviewed (approved):
Great job here, I think this is good to merge. I'm still hitting some bugs here and there (not sure what's a backend issue + what's the frontend), but nothing major, so I think we can merge this behind a FF, and then iron out the rest from there.

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-filter-logic

alexkim205 committed:
fix bug with values not being purged on value change

alexkim205 committed:
Merge remote-tracking branch 'origin' into feat/cohort-filter-logic

alexkim205 committed:
remove console
